python/lsst/daf/butler/core/datastore.py

```diff
@@ -47,6 +48,14 @@ class DatastoreValidationError(ValidationError):
     pass
 
 
+@dataclass(frozen=True)
+class Event:
+    name: str
+    undoFunc: typing.Callable
+    args: tuple
+    kwargs: dict
```

Comment:
I'd recommend using `__slots__` here, because there may be a lot of these.  (It's a real shame there's no way for `dataclass` to do that for you, but I've looked into it - apparently the decorator fires too late for anything it does in that department to matter).

Reply:
It had not even crossed my mind that dataclasses did not use slots. I would have put money on them being specifically designed to be efficient.

I see much discussion at ericvsmith/dataclasses#28